nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java

```diff
@@ -0,0 +1,502 @@
+            case ARRAY: {
+                final DataType arrayDataType = ((ArrayDataType) dataType).getElementType();
+
+                final Object newValue = parseFieldForType(startElement, fieldName, arrayDataType, recordValues);
+                final Object oldValues = recordValues.get(fieldName);
+
+                if (newValue != null) {
+                    if (oldValues != null) {
+                        if (oldValues instanceof List) {
+                            ((List) oldValues).add(newValue);
+                        } else {
+                            return new ArrayList<Object>(){{ add(oldValues); add(newValue); }};
+                        }
+                    } else {
+                        return new ArrayList<Object>(){{ add(newValue); }};
+                    }
+                } else {
+                    return null;
+                }
+            }
+            case RECORD: {
+                final RecordSchema childSchema;
+                if (dataType instanceof RecordDataType) {
+                    childSchema = ((RecordDataType) dataType).getChildSchema();
+                } else {
+                    return null;
+                }
+                return parseRecord(startElement, childSchema, true, true);
+            }
+            case MAP: {
+                logger.warn("Type map is not supported by this record reader. Field will be skipped.");
+                skipElement();
+                return null;
+            }
+            case CHOICE: {
+                return parseUnknownField(startElement);
+            }
+        }
+        return null;
+    }
+
+    private Object parseUnknownField(StartElement startElement) throws XMLStreamException {
+        // parse attributes
+        final Map<String, Object> recordValues = new HashMap<>();
+        final Iterator iterator = startElement.getAttributes();
+        while (iterator.hasNext()) {
+            final Attribute attribute = (Attribute) iterator.next();
+            final String attributeName = attribute.getName().toString();
+            recordValues.put(attributePrefix == null ? attributeName : attributePrefix + attributeName, attribute.getValue());
+        }
+        boolean hasAttributes = recordValues.size() > 0;
+
+        // parse fields
+        while (xmlEventReader.hasNext()) {
+            final XMLEvent xmlEvent = xmlEventReader.nextEvent();
+            if (xmlEvent.isCharacters()) {
+                final Characters characters = xmlEvent.asCharacters();
+                if (!characters.isWhiteSpace()) {
+                    xmlEventReader.nextEvent();
+                    if (hasAttributes) {
+                        if (contentFieldName != null) {
+                            recordValues.put(contentFieldName, characters.toString());
+                        } else {
+                            logger.debug("Found content for field that has to be parsed as record but property \"Field Name for Content\" is not set. " +

```

Comment:
I wonder if it makes sense in this case to use the name of the field  as the 'contentFieldName'. For example, if we had the following XML snippet:
```
<person>
  <location gurantee="NONE">1.469, -3.875</location>
</person>
```
Then if the Content Field Name were null, we would just use the name of the field (in this case 'location') so that a flattened representation of the record would look like:
```
person.location.guarantee = NONE
person.location.location = 1.469, -3.875
```
Thoughts?

Reply:
nice idea

Reply:
But what if the xml has the following structure?

```
<person>
  <location location="NONE">1.469, -3.875</location>
</person>
```

then the will replace the attribute as long as location ist not defined as an array...